### PR TITLE
Add `fragment_query` generic extractor arg for DASH and native HLS

### PR DESCRIPTION
**Pull request by bashonly**
Some DASH and HLS streams require the query from the manifest URL to be appended to its segment/fragment URLs as well, even if the segment/fragment URLs listed in the media playlist or manifest do not have any query.

A couple examples:
https://github.com/yt-dlp/yt-dlp/issues/4786#issuecomment-1296787075
https://github.com/yt-dlp/yt-dlp/pull/5604

This PR:
 - Adds support to the DASH downloader for query passthrough to fragments via the info_dict field `extra_param_to_segment_url`(the native HLS downloader already supported this)
 - Adds an extractor arg for the GenericIE, `fragment_query`, which when passed will add the manifest URL's query to the info_dict's `extra_param_to_segment_url` field

Usage: `yt-dlp --extractor-args generic:fragment_query "GENERIC_M3U8_OR_MPD_URL"`

Unfortunately this will not work with ffmpeg.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/generic.py [hunk 1]
```diff
@@ -2608,6 +2617,8 @@ def _real_extract(self, url):
             subtitles = {}
             if format_id.endswith('mpegurl'):
                 formats, subtitles = self._extract_m3u8_formats_and_subtitles(url, video_id, 'mp4', headers=headers)
+                if self._hls_segment_query(url):
+                    info_dict.update(self._hls_segment_query(url))
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:2621]:
```suggestion
                info_dict.update(self._hls_segment_query(url))
```
same everywhere

File: README.md [hunk 2]
```diff
@@ -1736,6 +1736,9 @@ The following extractors use this feature:
 * `skip`: One or more of `webpage` (skip initial webpage download), `authcheck` (allow the download of playlists requiring authentication when no initial webpage is downloaded. This may cause unwanted behavior, see [#1122](https://github.com/yt-dlp/yt-dlp/pull/1122) for more details)
 * `approximate_date`: Extract approximate `upload_date` and `timestamp` in flat-playlist. This may cause date-based filters to be slightly off
 
+#### generic
+* `hls_segment_query`: If a generic m3u8 URL has a query, have the native HLS downloader append the query to the HLS segment URLs. Does not apply to ffmpeg
```
Comment (pukkandan) [inline README.md:1740]:
```suggestion
* `hls_segment_query`: Passthrough any query in mpd/m3u8 manifest URLs to their fragments. Does not apply to ffmpeg
```
See https://github.com/yt-dlp/yt-dlp/pull/5604#discussion_r1027252728

File: yt_dlp/extractor/generic.py [hunk 3]
```diff
@@ -2446,6 +2446,15 @@ def report_detected(self, name, num=1, note=None):
...
         self._downloader.write_debug(f'Identified {num} {name}{format_field(note, None, "; %s")}')
 
+    def _hls_segment_query(self, url):
+        if self._configuration_arg('hls_segment_query', ie_key=GenericIE):
+            query_string = urllib.parse.urlparse(url).query
+            if query_string:
+                return {
+                    'extra_param_to_segment_url': query_string,
+                }
+        return
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:2456]:
```suggestion
        if self._configuration_arg('hls_segment_query'):
            query_string = urllib.parse.urlparse(url).query
            if query_string:
                return {'extra_param_to_segment_url': query_string}
        return {}
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (bashonly) [reply README.md:1740]: should this be renamed to just `segment_query` or `fragment_query` if it's going to be the same arg used for HLS and DASH?

Comment (pukkandan) [reply README.md:1740]: Pls do

Comment (bashonly) [reply README.md:1740]: also, when working on #5604, I encountered HLS media playlist URLs that required their master playlist's URL query passed to them. should I add functionality for that in this PR as well? and if so, should it be triggered by the same arg, or a different one, like `variant_query` or something

Comment (pukkandan) [reply README.md:1740]: hm.. Since we don't have any generic example of this, I wouldn't bother. But if you want to add it, I won't stop you either

**Review by pukkandan — APPROVED** [review]

**Pull request merged**